### PR TITLE
HTTPCLIENT-2222: fixed ProtocolNegotiationException

### PR DESCRIPTION
**Pull request by gaojieliu**
Aligned with the implementation of the latest release to get rid of
ProtocolNegotiationException:
https://github.com/apache/httpcomponents-core/blob/rel/v5.2-beta2/httpcore5-h2/src/main/java/org/apache/hc/core5/http2/impl/nio/ClientH2PrefaceHandler.java#L142

cc @ok2c

### CODE REVIEW

File: httpcore5-h2/src/main/java/org/apache/hc/core5/http2/impl/nio/H2OnlyClientProtocolNegotiator.java [hunk 1]
```diff
@@ -104,7 +106,8 @@ private void writeOutPreface(final IOSession session) throws IOException  {
         }
         if (!preface.hasRemaining()) {
             session.clearEvent(SelectionKey.OP_WRITE);
-            startProtocol(HttpVersion.HTTP_2, new ClientH2IOEventHandler(http2StreamHandlerFactory.create(ioSession)), null);
+            final ByteBuffer data = inBuf != null ? inBuf.data() : null;
```
Comment (carterkozak) [inline httpcore5-h2/src/main/java/org/apache/hc/core5/http2/impl/nio/H2OnlyClientProtocolNegotiator.java:109]: We can avoid the unnecessary additional volatile read (I find this pattern helpful in general, especially when null checks are involved, though I realize in this case we never transition from non-null to null)
```suggestion
            final BufferedData inBufSnapshot = inBuf;
            final ByteBuffer data = inBufSnapshot != null ? inBufSnapshot.data() : null;
```

File: httpcore5-h2/src/main/java/org/apache/hc/core5/http2/impl/nio/H2OnlyClientProtocolNegotiator.java [hunk 2]
```diff
@@ -131,7 +134,10 @@ public void outputReady(final IOSession session) throws IOException {
     @Override
     public void inputReady(final IOSession session, final ByteBuffer src) throws IOException {
         if (src != null) {
-            throw new ProtocolNegotiationException("Unexpected input");
+            if (inBuf == null) {
+                inBuf = BufferedData.allocate(src.remaining());
+            }
+            inBuf.put(src);
```
Comment (carterkozak) [inline httpcore5-h2/src/main/java/org/apache/hc/core5/http2/impl/nio/H2OnlyClientProtocolNegotiator.java:140]: I think we need to handle a potential buffer overflow here in the case an existing `inBuf` doesn't have enough room for `src`

File: httpcore5-h2/src/main/java/org/apache/hc/core5/http2/impl/nio/H2OnlyClientProtocolNegotiator.java (hunk 1 above)
Comment (ok2c) [reply httpcore5-h2/src/main/java/org/apache/hc/core5/http2/impl/nio/H2OnlyClientProtocolNegotiator.java:109]: @carterkozak `IOEventHandler`s can be called by a single I/O event dispatch thread at a time by design and therefore do not need to synchronize access to their internal structures if they do not delegate event handling to external handlers.

**Comment by ok2c** [comment]
@gaojieliu These are reasonable changes, though, I thought they were only relevant to protocol negotiators in `master` that have been completely rewritten in the course of 5.2 development.

**Comment by gaojieliu** [comment]
Hi @ok2c 
Thanks for the update, and enjoy your vacation!

**Comment by gaojieliu** [comment]
Hi @garydgregory 
Thank a lot, and let me give it a try!

**Pull request merged**